File: src/main/java/org/jukeboxmc/block/behavior/BlockWoodenSlab.java
package org.jukeboxmc.block.behavior;

import com.nukkitx.nbt.NbtMap;
import org.jetbrains.annotations.NotNull;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.BlockType;
import org.jukeboxmc.block.data.WoodType;
import org.jukeboxmc.block.direction.BlockFace;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.item.ItemType;
import org.jukeboxmc.item.behavior.ItemWoodenSlab;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.util.Identifier;
import org.jukeboxmc.world.World;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockWoodenSlab extends BlockSlab {

    public BlockWoodenSlab( Identifier identifier ) {
        super( identifier );
    }

    public BlockWoodenSlab( Identifier identifier, NbtMap blockStates ) {
        super( identifier, blockStates );
    }

    @Override
    public boolean placeBlock(@NotNull Player player, @NotNull World world, Vector blockPosition, @NotNull Vector placePosition, @NotNull Vector clickedPosition, Item itemInHand, BlockFace blockFace ) {
        Block targetBlock = world.getBlock( blockPosition );
        Block block = world.getBlock( placePosition );

        if ( blockFace == BlockFace.DOWN ) {
            if ( targetBlock instanceof BlockWoodenSlab && ( (BlockWoodenSlab) targetBlock ).isTopSlot() && ( (BlockWoodenSlab) targetBlock ).getWoodType().equals( this.getWoodType() ) ) {
                world.setBlock( blockPosition, Block.<BlockDoubleWoodenSlab>create( BlockType.DOUBLE_WOODEN_SLAB ).setWoodType( this.getWoodType() ) );
                return true;
            } else if ( block instanceof BlockWoodenSlab && ( (BlockWoodenSlab) block ).getWoodType().equals( this.getWoodType() ) ) {
                world.setBlock( placePosition, Block.<BlockDoubleWoodenSlab>create( BlockType.DOUBLE_WOODEN_SLAB ).setWoodType( this.getWoodType() ) );
                return true;
            }
        } else if ( blockFace == BlockFace.UP ) {
            if ( targetBlock instanceof BlockWoodenSlab && !( (BlockWoodenSlab) targetBlock ).isTopSlot() && ( (BlockWoodenSlab) targetBlock ).getWoodType().equals( this.getWoodType() ) ) {
                world.setBlock( blockPosition, Block.<BlockDoubleWoodenSlab>create( BlockType.DOUBLE_WOODEN_SLAB ).setWoodType( this.getWoodType() ) );
                return true;
            } else if ( block instanceof BlockWoodenSlab && ( (BlockWoodenSlab) block ).getWoodType().equals( this.getWoodType() ) ) {
                world.setBlock( placePosition, Block.<BlockDoubleWoodenSlab>create( BlockType.DOUBLE_WOODEN_SLAB ).setWoodType( this.getWoodType() ) );
                return true;
            }
        } else {
            if ( block instanceof BlockWoodenSlab && ( (BlockWoodenSlab) block ).getWoodType().equals( this.getWoodType() ) ) {
                world.setBlock( placePosition, Block.<BlockDoubleWoodenSlab>create( BlockType.DOUBLE_WOODEN_SLAB ).setWoodType( this.getWoodType() ) );
                return true;
            } else {
                this.setTopSlot( clickedPosition.getY() > 0.5 && !world.getBlock( blockPosition ).canBeReplaced( this ) );
            }
        }
        super.placeBlock( player, world, blockPosition, placePosition, clickedPosition, itemInHand, blockFace );
        world.setBlock( placePosition, this );
        return true;
    }

    @Override
    public Item toItem() {
        return Item.<ItemWoodenSlab>create( ItemType.WOODEN_SLAB ).setWoodType( this.getWoodType() );
    }

    public BlockWoodenSlab setWoodType(@NotNull WoodType woodType ) {
        return this.setState( "wood_type", woodType.name().toLowerCase() );
    }

    public @NotNull WoodType getWoodType() {
        return this.stateExists( "wood_type" ) ? WoodType.valueOf( this.getStringState( "wood_type" ) ) : WoodType.OAK;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockWoodenSlab.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.Block
import org.jukeboxmc.api.block.BlockType
import org.jukeboxmc.api.block.WoodenSlab
import org.jukeboxmc.api.block.data.BlockFace
import org.jukeboxmc.api.block.data.VerticalHalf
import org.jukeboxmc.api.item.Item
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.server.block.JukeboxBlock
import org.jukeboxmc.server.extensions.toJukeboxBlock
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer
import org.jukeboxmc.server.world.JukeboxWorld

class BlockWoodenSlab(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates), WoodenSlab {

    override fun placeBlock(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        placePosition: Vector,
        clickedPosition: Vector,
        itemInHand: JukeboxItem,
        blockFace: BlockFace
    ): Boolean {
        val targetBlock = world.getBlock(blockPosition)
        val block = world.getBlock(placePosition)

        val doubleWoodenType = when(this.getType()) {
            BlockType.OAK_SLAB -> BlockType.OAK_DOUBLE_SLAB
            BlockType.SPRUCE_SLAB -> BlockType.SPRUCE_DOUBLE_SLAB
            BlockType.BIRCH_SLAB -> BlockType.BIRCH_DOUBLE_SLAB
            BlockType.JUNGLE_SLAB -> BlockType.JUNGLE_DOUBLE_SLAB
            BlockType.ACACIA_SLAB -> BlockType.ACACIA_DOUBLE_SLAB
            BlockType.DARK_OAK_SLAB -> BlockType.DARK_OAK_DOUBLE_SLAB
            BlockType.MANGROVE_SLAB -> BlockType.MANGROVE_DOUBLE_SLAB
            BlockType.CHERRY_SLAB -> BlockType.CHERRY_DOUBLE_SLAB
            BlockType.BAMBOO_SLAB -> BlockType.BAMBOO_DOUBLE_SLAB
            BlockType.BAMBOO_MOSAIC_SLAB -> BlockType.BAMBOO_MOSAIC_DOUBLE_SLAB
            BlockType.CRIMSON_SLAB -> BlockType.CRIMSON_DOUBLE_SLAB
            BlockType.WARPED_SLAB -> BlockType.WARPED_DOUBLE_SLAB
            else -> BlockType.AIR
        }

        if (block is BlockWater && block.getLiquidDepth() == 0) {
            world.setBlock(placePosition, block, 1, false)
        }

        if (blockFace === BlockFace.DOWN) {
            if (targetBlock is BlockWoodenSlab && targetBlock.getVerticalHalf() == VerticalHalf.TOP && this.getWoodType(targetBlock.getType()) == this.getWoodType(this.getType())) {
                world.setBlock(blockPosition, Block.create<BlockDoubleWoodenSlab>(doubleWoodenType))
                return true
            } else if (block is BlockWoodenSlab && this.getWoodType(block.getType()) == this.getWoodType(this.getType())) {
                world.setBlock(placePosition, Block.create<BlockDoubleWoodenSlab>(doubleWoodenType))
                return true
            }
        } else if (blockFace === BlockFace.UP) {
            if (targetBlock is BlockWoodenSlab && targetBlock.getVerticalHalf() != VerticalHalf.TOP && this.getWoodType(targetBlock.getType()) == this.getWoodType(this.getType())) {
                world.setBlock(blockPosition, Block.create<BlockDoubleWoodenSlab>(doubleWoodenType))
                return true
            } else if (block is BlockWoodenSlab && this.getWoodType(block.getType()) == this.getWoodType(this.getType())) {
                world.setBlock(placePosition, Block.create<BlockDoubleWoodenSlab>(doubleWoodenType))
                return true
            }
        } else {
            if (block is BlockWoodenSlab && this.getWoodType(block.getType()) == this.getWoodType(this.getType())) {
                world.setBlock(placePosition, Block.create<BlockDoubleWoodenSlab>(doubleWoodenType))
                return true
            } else {
                this.setVerticalHalf(if (clickedPosition.getY() > 0.5 && !world.getBlock(blockPosition).toJukeboxBlock().canBeReplaced(this)) VerticalHalf.TOP else VerticalHalf.BOTTOM)
            }
        }
        return super.placeBlock(player, world, blockPosition, placePosition, clickedPosition, itemInHand, blockFace)
    }

    override fun getVerticalHalf(): VerticalHalf {
        return VerticalHalf.valueOf(this.getStringState("minecraft:vertical_half"))
    }

    override fun setVerticalHalf(value: VerticalHalf): WoodenSlab {
        return this.setState("minecraft:vertical_half", value.name.lowercase())
    }

    override fun getDrops(item: Item): MutableList<Item> {
        return this.createItemDrop(item, this.toItem())
    }

    override fun getWaterLoggingLevel(): Int {
        return 1
    }

    private fun getWoodType(type: BlockType): String {
        return when (type) {
            BlockType.OAK_SLAB, BlockType.OAK_DOUBLE_SLAB -> "oak"
            BlockType.SPRUCE_SLAB, BlockType.SPRUCE_DOUBLE_SLAB -> "spruce"
            BlockType.BIRCH_SLAB, BlockType.BIRCH_DOUBLE_SLAB -> "birch"
            BlockType.JUNGLE_SLAB, BlockType.JUNGLE_DOUBLE_SLAB -> "jungle"
            BlockType.ACACIA_SLAB, BlockType.ACACIA_DOUBLE_SLAB -> "acacia"
            BlockType.DARK_OAK_SLAB, BlockType.DARK_OAK_DOUBLE_SLAB -> "dark_oak"
            BlockType.MANGROVE_SLAB, BlockType.MANGROVE_DOUBLE_SLAB -> "mangrove"
            BlockType.CHERRY_SLAB, BlockType.CHERRY_DOUBLE_SLAB -> "cherry"
            BlockType.BAMBOO_SLAB, BlockType.BAMBOO_DOUBLE_SLAB -> "bamboo"
            BlockType.BAMBOO_MOSAIC_SLAB, BlockType.BAMBOO_MOSAIC_DOUBLE_SLAB -> "bamboo_mosaic"
            BlockType.CRIMSON_SLAB, BlockType.CRIMSON_DOUBLE_SLAB -> "crimson"
            BlockType.WARPED_SLAB, BlockType.WARPED_DOUBLE_SLAB -> "warped"
            else -> ""
        }
    }
}